src/test/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReaderTest.java

```diff
@@ -0,0 +1,143 @@
+ * the License.
+ */
+
+package io.cdap.plugin.servicenow.source;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.macro.Macros;
+import io.cdap.cdap.api.plugin.PluginProperties;
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.internal.AssumptionViolatedException;
+import org.junit.rules.ExpectedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+
+public class ServiceNowRecordReaderTest {
+
+  private static final String CLIENT_ID = System.getProperty("servicenow.test.clientId");
+  private static final String CLIENT_SECRET = System.getProperty("servicenow.test.clientSecret");
+  private static final String REST_API_ENDPOINT = System.getProperty("servicenow.test.restApiEndpoint");
+  private static final String USER = System.getProperty("servicenow.test.user");
+  private static final String PASSWORD = System.getProperty("servicenow.test.password");
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowMultiInputFormatTest.class);
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+  private ServiceNowSourceConfig serviceNowSourceConfig;
+  private ServiceNowRecordReader serviceNowRecordReader;
+
+  @Before
+  public void initializeTests() {
+    try {
+      Assume.assumeNotNull(CLIENT_ID, CLIENT_SECRET, REST_API_ENDPOINT, USER, PASSWORD);
+      serviceNowSourceConfig = ServiceNowSourceConfigHelper.newConfigBuilder()
+        .setReferenceName("referenceName")
+        .setRestApiEndpoint(REST_API_ENDPOINT)
+        .setUser(USER)
+        .setPassword(PASSWORD)
+        .setClientId(CLIENT_ID)
+        .setClientSecret(CLIENT_SECRET)
+        .setTableNames("sys_user")
+        .setValueType("Actual")
+        .setStartDate("2021-12-30")
+        .setEndDate("2021-12-31")
+        .setTableNameField("tablename")
+        .build();
+
+      serviceNowRecordReader = new ServiceNowRecordReader(serviceNowSourceConfig);
+
+    } catch (AssumptionViolatedException e) {
+      LOG.warn("Service Now batch multi source tests are skipped. ");
+      throw e;
+    }
+  }
+
+  @Test
+  public void testConstructor() throws IOException {
+    serviceNowRecordReader.close();
+    Assert.assertEquals(0, serviceNowRecordReader.pos);
+  }
+
+  @Test
+  public void testConstructor2() throws IOException {
+
+    serviceNowRecordReader.close();
+    Assert.assertEquals(0, serviceNowRecordReader.pos);
+    Assert.assertEquals("pipeline.user.1", serviceNowSourceConfig.getUser());
+    Assert.assertEquals("tablename", serviceNowSourceConfig.getTableNameField());
+    Assert.assertEquals("tablename", serviceNowSourceConfig.getTableName());
+    Assert.assertEquals("2021-12-30", serviceNowSourceConfig.getStartDate());
+    Assert.assertEquals("https://ven05127.service-now.com", serviceNowSourceConfig.getRestApiEndpoint());
```

Comment:
do these tests only work against a specific salesforce instance (i.e. your instance with your credentials)?

Reply:
They'll work against all Service Now instances. The following Assertions:

- `Assert.assertEquals("pipeline.user.1", serviceNowSourceConfig.getUser());`
- `Assert.assertEquals("https://ven05127.service-now.com", serviceNowSourceConfig.getRestApiEndpoint());` 
are dependent on the the user provided VM arguments while running the test case. If the system properties such has _CLIENT_ID, CLIENT_SECRET, REST_API_ENDPOINT, USER, PASSWORD_  are not provided in the VM arguments during test case run, then the test case gets ignored.

If you'd like, I can have these 2 assertions removed.

Reply:
I've removed all test cases and assertions that were dependent upon user credentials.